cmd/registry/scoring/parser.go

```diff
@@ -0,0 +1,153 @@
+				errs := validateScoreFormula(targetName, scoreFormula)
+				formulaErrs = append(formulaErrs, errs...)
+			}
+		}
+	}
+
+	thresholdErrs := make([]error, 0)
+	// Validate threshold
+	switch scoreType := scoreDefinition.GetType().(type) {
+	case *rpc.ScoreDefinition_Percent:
+		// minValue: 0 maxValue:100
+		// validate that the set thresholds are within these bounds
+		errs := validateNumberThresholds(scoreType.Percent.GetThresholds(), 0, 100)
+		thresholdErrs = append(thresholdErrs, errs...)
+	case *rpc.ScoreDefinition_Integer:
+		// defaults if not set: minValue: 0 maxValue:0
+		// validate that the set thresholds are within these bound
+		errs := validateNumberThresholds(scoreType.Integer.GetThresholds(), scoreType.Integer.GetMinValue(), scoreType.Integer.GetMaxValue())
+		thresholdErrs = append(thresholdErrs, errs...)
+	case *rpc.ScoreDefinition_Boolean:
+		errs := validateBooleanThresholds(scoreType.Boolean.GetThresholds())
+		thresholdErrs = append(thresholdErrs, errs...)
+	}
+
+	errs := make([]error, len(patternErrs)+len(formulaErrs)+len(thresholdErrs))
+	errs = append(errs, patternErrs...)
+	errs = append(errs, formulaErrs...)
+	errs = append(errs, thresholdErrs...)
+
+	return errs
+}
+
+func validateScoreFormula(targetName patterns.ResourceName, scoreFormula *rpc.ScoreFormula) []error {
+	errs := make([]error, 0)
+
+	// Validation checks for score_formula.artifact.pattern
+	pattern := scoreFormula.GetArtifact().GetPattern()
+
+	// pattern should have a valid $resource pattern
+	_, entityType, err := patterns.GetReferenceEntityType(pattern)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("invalid score_formula.artifact.pattern: %s", err))
+	}
+	// $resource should have valid entity reference wrt target_resource
+	_, err = patterns.GetReferenceEntityValue(pattern, targetName)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("invalid $resource reference in score_formula.artifact.pattern: %s", err))
+	}
+	// pattern should always start with a $resource reference
+	if entityType == "default" {
+		errs = append(errs, fmt.Errorf("invalid score_formula.artifact.pattern: %q, must always start with '$resource.(api|version|spec|artifact)'", pattern))
+	}
+	// score_formula.pattern.pattern should not end with a "-"
+	if strings.HasSuffix(scoreFormula.GetArtifact().GetPattern(), "/-") {
+		errs = append(errs, fmt.Errorf("invalid score_formula.artifact.pattern : %q, it should end with a resourceID and not a \"-\"", pattern))
+	}
+
+	return errs
+}
+
+func validateNumberThresholds(thresholds []*rpc.NumberThreshold, minValue, maxValue int32) []error {
+	errs := make([]error, 0)
+
+	// sort the thresholds based on range.min
+	sort.Slice(thresholds, func(i, j int) bool {
+		return thresholds[i].GetRange().GetMin() < thresholds[j].GetRange().GetMin()
+	})
+
+	// This loop runs a pointer through the sorted ranges and checks for overlaps, missed values or out of bound scenarios.
```

Comment:
I'm having some trouble following the algorithm. Is it possible to simplify?

Something like this might work.
1. Check if any of the thresholds have nil ranges. Do we need this step?
2. Sort by range.min
3. Assert the lowest threshold minimum == minValue
4. For each pair of thresholds in the list, assert that left.Max == right.Min - 1
```
for i := 1; i < len(thresholds); i++ {
	left, right := thresholds[i - 1].GetRange(), thresholds[i].GetRange()
	if left.GetMax() < right.GetMin() - 1 {
		// Handle gap.
	} else if left.GetMax() > right.GetMin() - 1 {
		// Handle overlap.
	}
}
```
5. Assert the highest threshold maximum == maxValue

Reply:
Done.